packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerParams.ts

```diff
@@ -0,0 +1,269 @@
+    const newRuntimeVals = runtimes.map(runtime => runtime.value).sort();
+    const areRuntimesSame = _.isEqual(existingRuntimeVals, newRuntimeVals);
+    if (!areRuntimesSame) {
+      this.updateRuntimes(runtimes);
+      if (this.isLatestVersionFinalized()) {
+        this.addNewLayerVersion();
+      }
+    }
+  }
+
+  setPermissionsForVersion(version: number, permissions: LayerPermission[]) {
+    this.storedParams.layerVersionMap[version].permissions = permissions;
+    this.versionMap.get(version).setPermissions(permissions);
+  }
+
+  toStoredLayerParameters(): StoredLayerParameters {
+    return _.cloneDeep(this.storedParams);
+  }
+
+  private isLatestVersionFinalized(): boolean {
+    return this.getHash(this.getLatestVersion()) !== undefined;
+  }
+
+  private hashLayer() {
+    const layerPath = path.join(this.context.amplify.pathManager.getBackendDirPath(), category, this.layerName);
+    return hashLayerVersionContents(layerPath);
+  }
+
+  private updateRuntimes(runtimes: LayerRuntime[]) {
+    this.runtimes = runtimes;
+    this.storedParams.runtimes = runtimes;
+  }
+
+  private addNewLayerVersion() {
+    const currVersion = this.getLatestVersion();
+    const newVersion = currVersion + 1;
+    const prevPermissions = this.getVersion(currVersion).permissions;
+    this.storedParams.layerVersionMap[newVersion] = {
+      permissions: _.cloneDeep(prevPermissions),
+    };
+    this.versionMap.set(newVersion, new LayerVersionState(this.storedParams.layerVersionMap[newVersion]));
+  }
+}
+
+class LayerVersionState implements LayerVersionMetadata {
+  permissions: LayerPermission[];
+  hash: string;
+  constructor(versionData: Pick<LayerVersionMetadata, 'permissions' | 'hash'>) {
+    this.permissions = [];
+    this.hash = versionData.hash;
+    this.setPermissions(versionData.permissions);
+  }
+
+  setPermissions(permissions: LayerPermission[]) {
+    this.permissions = [];
+    permissions.forEach(permission => {
+      if (permission.type === Permission.public) {
+        const permissionPublic: PublicLayer = {
+          type: Permission.public,
+        };
+        this.permissions.push(permissionPublic);
+      }
+      if (permission.type === Permission.private) {
+        const permissionPrivate: PrivateLayer = {
+          type: Permission.private,
+        };
+
+        this.permissions.push(permissionPrivate);
+      }
+      if (permission.type === Permission.awsOrg) {
+        const orgsPermissions: OrgsLayer = {
+          type: Permission.awsOrg,
+          orgs: permission.orgs,
+        };
+        this.permissions.push(orgsPermissions);
+      }
+      if (permission.type === Permission.awsAccounts) {
+        const accountsPermissions: AccountsLayer = {
+          type: Permission.awsAccounts,
+          accounts: permission.accounts,
+        };
+        this.permissions.push(accountsPermissions);
+      }
+    });
+  }
+
+  listAccoutAccess(): string[] {
```

Comment:
`listAccountAccess` typo.

Also what is the expected behavior? In case of multiple Permission.awsAccounts it will return only the last one and not appending. Why not filter on the permissions? Same for below, why do we need functions for these?

Reply:
there will only be one instance of each type (the single instance stores a list of all the awsAccounts). Will change filter to find to make that clear.